DOCS-8072-4.3-tuning-recommendations-kt (#1408)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/tuning-recommendations.adoc

= Tuning Recommendations
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Mule 4 runtime engine provides self-tuning execution, which adjusts itself for optimal performance according to the underlying operating conditions of the environment where you deploy Mule runtime engine (Mule). +

Reviewer: ```suggestion
Mule 4 runtime engine comes to you already tuned for optimal performance for most common use cases. Additionally, it includes self-tuning execution, which adjusts itself for optimal performance according to the underlying operating conditions of the environment where you deploy Mule runtime engine (Mule). This self-tuning feature frees you to focus on the specifics of your application and apply tuning recommendations at that level only.+
```
I recommend not even mentioning changing Mule 4.

Self-tuning allows you to focus on the specifics of your application rather than Mule internals, which eases the optimization process. Besides, Mule 4 out of the box configuration is already tuned for optimal performance for most common use cases, and changing any parameter is advised against unless you fully understand the implications of doing so.

> There are tuning recommendations that you can apply at the application level and benefit regardless of your use case:

Reviewer: See previous edit, incorporating this text.

However, I have a question about the final sentence. Do you intend it to point to the subtopics? We should not end a topic with a colon; it makes me wonder if the following subtopics should actually be LIST ITEMS, as I suggested in PR 1407.

In that case, this could all be one big topic...:

"...This self-tuning feature of Mule frees you to tune your application only, based on the following knowledge: 
* Streaming strategy +
<text>
* Back-pressure and MaxConcurrency  +
<text>
* Back-end server response time
<text>
.
.
.

The following are pretty brief to be considered actual topics; they sound like a list of To-Do items rather than concepts. Copyedited only.

> == Repeatable vs Non-repeatable Streaming

Reviewer: ```suggestion
== Repeatable vs. Nonrepeatable Streaming
```

> Understand which streaming strategy applies on your use case to get the best performance of your flows: +

Reviewer: ```suggestion
Understand which streaming strategy results in the best flow performance for your use case: +
```

* A repeatable stream (read payload more than once)

> * A non-repeatable stream (read payload only once)

Reviewer: ```suggestion
* A nonrepeatable stream (read payload only once)
```

See xref:tuning-repeatable-nonrepeatable.adoc[Repeatable vs Non-repeatable Streaming] for details.

> == Back-pressure and Max Concurrency

Reviewer: ```suggestion
== Back-Pressure and MaxConcurrency
```

> Understand the back-pressure messages that Mule logs when it can not handle and event and learn how to use the `maxConcurrency` parameter to set the maximum number of concurrent messages that a flow can process at any given time.

Reviewer: ```suggestion
Understand back-pressure and learn how to use the `maxConcurrency` parameter to tune the number of concurrent messages sent to your flow.
```

See xref:tuning-backpressure-maxconcurrency.adoc[Back-Pressure and MaxConcurrency] for details.

> == Backend Servers Response Time

Reviewer: ```suggestion
== Backend Server Response Time
```

> Understand if the backend servers average latency and throughput are limiting the scalability or performance of your application.

Reviewer: ```suggestion
Understand if the average latency and throughput of backend servers is limiting the scalability or performance of your application.
```

See xref:tuning-backend-server.adoc[Backend Server Response Time] for details.

== Caching

> Identifying key aspects of your data leads to a better solution approach. Mule offers some customizable mechanisms, such as Cache Scope and HTTP Caching API Gateway policy, to enable cache according to your needs. Understand when to use cache and which cache-strategy to follow.

Reviewer: parallelism
```suggestion
Understand when to use cache and which cache-strategy to follow, based on key aspects of your data. Mule offers some customizable mechanisms, such as Cache Scope and the HTTP Caching API Gateway policy, to enable cache according to your needs.
```

See xref:tuning-caching.adoc[Caching] for details.

> 

Reviewer: Ah, OK: I just got your Slack about this being the work of multiple people, so I won't fixate on organization and will just make sure it all sounds right. Thanks for setting me straight! 
At some point in the future, perhaps we can look at the org.  :-)

Reviewer: For parallelism, consider rewording each subtopic to begin with "Understand."

== Pooling Profiles

> Pooled components configure their component pool which contains multiple instances of the component to handle simultaneous incoming requests. Understand how to add add `pooling-profile` to connectors when performance tests show it necessary.

Reviewer: ```suggestion
Pooling components helps process simultaneous incoming requests. Understand how to add add `pooling-profile` to connectors when performance tests show that it is necessary.
```

See xref:tuning-pooling-profiles.adoc[Pooling Profiles] for details.

== Domains

> By having a centralized-point for all the shared resources, domains help the class-loading process. Domains are also a performance choice when deploying multiple services on the same on-premises Mule runtime engine.

Reviewer: ```suggestion
Using domains provides a central repository for all the shared resources, facilitating the class-loading process. Domains enhance performance when you deploy multiple services on the same on-premises instance of Mule runtime engine.
```

See xref:tuning-domains.adoc[Domains] for details.

== Logging

> Understand why asynchronous logging has better performance  achieved than in synchronous logging strategy.

Reviewer: ```suggestion
Understand why asynchronous logging performs better than synchronous logging.
```

See xref:tuning-logging.adoc[Logging] for details.

== Batch Processing

> Mule possesses the ability to process messages in batches. Batch processing requires having enough available memory to process the threads in parallel, which means moving the records from persistent storage into RAM in a fixed size block. Understand how to configure the batch block size property for your application.

Reviewer: ```suggestion
Mule can process messages in batches, but batch processing requires having enough available memory to process the threads in parallel, which means moving the records from persistent storage into RAM in a fixed-size block. Understand how to configure the batch block size property for your application.
```

See xref:tuning-batch-processing.adoc[Batch Processing] for details.

== Application Design Best Practices

> These practices can help you on the application design phase to get a better performance for your Mule apps.

Reviewer: ```suggestion
Following certain practices at the design phase helps you to achieve better performance from your Mule apps after they are deployed.
```
